### ISSUE TITLE
Why Effective-FFT's results inconsistent with actual results?

### ISSUE THREAD

**Issue by ZeroE04** [NVIDIA/FasterTransformer#168]
Why does the code in the sample/cpp/encoder_sample.cc give inconsistent results when setting is_remove_padding equal to true and false for the same input?
Comment: 'is_remove_padding' =False,  the result is the same as the actual result.

such as print_to_file(encoder_param.transformer_out):
actual result: -0.222001, 1.049087, -0.278886,-0.451816 ...
is_remove_padding = False: same as actual result,
is_remove_padding = Ture: 1.979067, 1.000596, 0.830792, 0.121805.

who can tell me why?... I'm very confused, please help me, thx!!

**Comment by byshiue** [comment]
Do you skip the comparison of padding part?

**Comment by ZeroE04** [comment]
> 

what means of "comparison of padding part"? （The entire output is different...）outputs is notI basically followed the source code( only the embedding/LN part is add before encoder). I'm just compare the **encoder_param.transformer_out** between True and False of the flag('is_remove_padding')；


The results  starts at the beginning of the pointer：
actual result: -0.222001, 1.049087, -0.278886,-0.451816 ...
is_remove_padding = False: same as actual result,
is_remove_padding = Ture: 1.979067, 1.000596, 0.830792, 0.121805...

In addition, even using source code for the experiment, the results are different too...

I comparing those by a wrong way？thx and look forward to your reply!

**Comment by ZeroE04** [comment]
> Do you skip the comparison of padding part?

> 

what means of "comparison of padding part"? （The entire output is different...）outputs is notI basically followed the source code( only the embedding/LN part is add before encoder). I'm just compare the **encoder_param.transformer_out** between True and False of the flag('is_remove_padding')；


The results  starts at the beginning of the pointer：
actual result: -0.222001, 1.049087, -0.278886,-0.451816 ...
is_remove_padding = False: same as actual result,
is_remove_padding = Ture: 1.979067, 1.000596, 0.830792, 0.121805...

In addition, even using source code for the experiment, the results are different too...

I comparing those by a wrong way？thx and look forward to your reply!

**Comment by ZeroE04** [comment]
> Do you skip the comparison of padding part?

① what means of "comparison of padding part"? （The all output is different...no fragment is consistent) 
② I basically followed the source code( only add  embedding/LN part before encoder). I'm just compare the encoder_param.transformer_out between True and False of the flag('is_remove_padding')；
③ In addition, even using source code for the experiment, the results are different too...

The results starts at the beginning of the pointer：
actual result: -0.222001, 1.049087, -0.278886,-0.451816 ...
is_remove_padding = False: same as actual result,
is_remove_padding = Ture: 1.979067, 1.000596, 0.830792, 0.121805...



I comparing those by a wrong way？thx and look forward to your reply!

**Comment by ZeroE04** [comment]
> Do you skip the comparison of padding part?

I guess you人means is the pad at the end of the sentence does not need to be compared, I know that, the problem is that no fragment is the same now, include the first token of the first sentence.

**Comment by ZeroE04** [comment]
> Do you skip the comparison of padding part?

I guess your means is all paddings at the end of the every sentence does not need to be compared, I know that, the problem is that no fragment is the same now, include the first token of the first sentence.

**Comment by byshiue** [comment]
I don't know which version of code you test, can you try the dev/v5.0_beta branch?

**Comment by byshiue** [comment]
Besides, if you want to verify the correctness, I suggest to use the TensorFlow example or PyTorch example. C++ example is used to demonstrate how to call FT in most time, and does not verify the correctness of example. 
The correctness of examples of TensorFlow and PyTorch are verified.

**Comment by ZeroE04** [comment]
> I don't know which version of code you test, can you try the dev/v5.0_beta branch?

my branch is main, and the recent log is:

`commit d71027c5bc9cf8db82454baebaaadd8289bfe5e8
Merge: dd4c071 1093489
Author: byshiue <bhsueh@nvidia.com>
Date:   Mon Nov 15 08:49:43 2021 +0800

    Merge pull request #164 from Shengr/bert_tf-profile
    
    update flags for profile_bert_inference

commit 10934895385c3787ed7e0bb02a4fe78d4ac3458e
Author: shengr <sean.ran@outlook.com>
Date:   Sun Nov 14 17:17:47 2021 +0800

    update flags for profile_bert_inference`


Even if compile **cpp/encoder_sample.cc** directly without making any changes, the results are inconsistent between true and false of remove_padding.


ok， i try to complie dev/v5.0_beta branch,  thx very much.

**Comment by ZeroE04** [comment]
> Besides, if you want to verify the correctness, I suggest to use the TensorFlow example or PyTorch example. C++ example is used to demonstrate how to call FT in most time, and does not verify the correctness of example. The correctness of examples of TensorFlow and PyTorch are verified.

when i complie the branch of dev/v5.0_beta, this error happend:
/FasterTransformer/src/fastertransformer/kernels/reduce_kernel_utils.cuh:19:38: fatal error: cooperative_grous/reduce.h: No such file or directory

i find this dir in cuda's include path, but nothing has be find...

**Comment by byshiue** [comment]
What's docker image do you use?

**Comment by ZeroE04** [comment]
> What's docker image do you use?

I'm using locally environment, not docker. 
I can complie the main branch in my environment, but can not complie dev/v5.0_beta.

**Comment by byshiue** [comment]
The file `cooperative_grous/reduce.h` is added in CUDA 11. Do you use CUDA 11? 
It is related a new feature in the latest code. 

If you cannot install the CUDA 11, you can try the tag `v5.0_beta_2021.09_tag`. This tag does not include the file.

**Comment by ZeroE04** [comment]
> The file `cooperative_grous/reduce.h` is added in CUDA 11. Do you use CUDA 11? It is related a new feature in the latest code.
> 
> If you cannot install the CUDA 11, you can try the tag `v5.0_beta_2021.09_tag`. This tag does not include the file.

ok, that' great, thx very much!

**Comment by ZeroE04** [comment]
> **[byshiue](/byshiue) ** c
hi, byshiue, the out of v5.0_beta_2021.09_tag is very perfect, and Exactly as expected. 
And to be sure, the main branch does have bugs for remove_padding. 

Thx  again！

**Comment by ZeroE04** [comment]
> The file `cooperative_grous/reduce.h` is added in CUDA 11. Do you use CUDA 11? It is related a new feature in the latest code.
> 
> If you cannot install the CUDA 11, you can try the tag `v5.0_beta_2021.09_tag`. This tag does not include the file.

hi, byshiue, the out of v5.0_beta_2021.09_tag is very perfect, and Exactly as expected. 
And to be sure, the main branch does have bugs for remove_padding. 

Thx  again！

**Comment by ZeroE04** [comment]
> The file `cooperative_grous/reduce.h` is added in CUDA 11. Do you use CUDA 11? It is related a new feature in the latest code.
> 
> If you cannot install the CUDA 11, you can try the tag `v5.0_beta_2021.09_tag`. This tag does not include the file.

hi, byshiue, I have a new puzzle....
What is the use of LayerNormType parameter(default post_layernorm) in bert.cc forword? In hugging face implement of bert by torch, it seem just match with post_layernorm, and not need pre_layernorm?

**Comment by byshiue** [comment]
We have many different transformer models. Some of them use pre_layernorm. 
For example, the encoder part of standard transformer in Attention is All you need.

**Comment by ZeroE04** [comment]
> We have many different transformer models. Some of them use pre_layernorm. For example, the encoder part of standard transformer in Attention is All you need.

Thx very much, wishing your life delectation!
